src/worker.cc

```diff
@@ -11,11 +11,24 @@ extern "C" {
   static PyObject *RayError;
 }
 
+inline WorkerServiceImpl::WorkerServiceImpl(const std::string& worker_address)
+  : worker_address_(worker_address), death_pending_() {
+  send_queue_.connect(worker_address_, false);
+}
+
 Status WorkerServiceImpl::ExecuteTask(ServerContext* context, const ExecuteTaskRequest* request, ExecuteTaskReply* reply) {
-  task_ = request->task(); // Copy task
+  if (this->death_pending_) {
+    reply->set_rejected(true);
+  }
+  task_ = std::unique_ptr<Task>(this->death_pending_ ? NULL : new Task(request->task())); // Copy task
   RAY_LOG(RAY_INFO, "invoked task " << request->task().name());
-  Task* taskptr = &task_;
-  send_queue_.send(&taskptr);
+  WorkerMessage message = { &task_ };
+  send_queue_.send(&message);
+  return Status::OK;
+}
+
+Status WorkerServiceImpl::Die(ServerContext* context, const DieRequest* request, DieReply* reply) {
+  this->death_pending_ = true;
   return Status::OK;
 }
```

Comment:
Good catch, I wasn't thinking and put the code in the wrong method. It should have been in `Die`; see new version.